### PR TITLE
udev-node: do not create stack directory for by-diskseq symlink

### LINKED ISSUE #34637: udev/pid1: loopback diskseq symlinks in /run/ are created for each run and kept around until a restart cycle ends leading to inode exhaustion
If a unit uses a loop device for any reason, a diskseq symlink is created in /run/udev/links/ for each invocation. If a unit fails and it has Restart= set, the symlinks are all kept around until a rate limit is hit, or the unit is manually stopped, and only then they are cleared. If a unit sets StartlimitIntervalSec=0 to disable rate limiting, and always fail, eventually /run/ runs out of inodes due to the symlinks.

To reproduce:

```
mkdir /var/foo/bar -p
mksquashfs /var/foo /var/foo.raw
cat <<EOF> /etc/systemd/system/test.service
[Unit]
StartLimitIntervalSec=0
[Service]
ExecStart=false
Restart=on-failure
MountImages=/var/foo.raw:/var/foo
EOF
```

Then start the test.service and observe the number of links grow indefinitely `ls /run/udev/links | grep diskseq | wc -l`

While this happens, udevadm monitor shows a stream of:

```
UDEV  [849.134231] change   /devices/virtual/block/loop4 (block)
UDEV  [849.187554] change   /devices/virtual/block/loop4 (block)
KERNEL[849.370486] change   /devices/virtual/block/loop3 (block)
UDEV  [849.425013] change   /devices/virtual/block/loop3 (block)
KERNEL[849.448361] change   /devices/virtual/block/loop4 (block)
UDEV  [849.486913] change   /devices/virtual/block/loop4 (block)
```

udevadm info on the loop device:

```
P: /devices/virtual/block/loop0
M: loop0
R: 0
J: b7:0
U: block
T: disk
D: b 7:0
N: loop0
L: 0
S: disk/by-diskseq/478
Q: 479
E: DEVPATH=/devices/virtual/block/loop0
E: DEVNAME=/dev/loop0
E: DEVTYPE=disk
E: DISKSEQ=479
E: MAJOR=7
E: MINOR=0
E: SUBSYSTEM=block
E: USEC_INITIALIZED=5176916
E: NVME_HOST_IFACE=none
E: SYSTEMD_READY=0
E: DEVLINKS=/dev/disk/by-diskseq/478
E: TAGS=:systemd:
E: CURRENT_TAGS=:systemd:
```

Stepping through with gdb, I pin pointed the exact moment when the new link is created, and it's when the loop_device_unref from here is called:

https://github.com/systemd/systemd/blob/598bb6fde46550ce020dd1d9853b316b766cb923/src/shared/dissect-image.c#L4040

more precisely when `d->fd` is closed here:

https://github.com/systemd/systemd/blob/598bb6fde46550ce020dd1d9853b316b766cb923/src/shared/loop-util.c#L812

It feels like we are keeping <some> reference somewhere that prevents udev from cleaning up these links while a restart cycle is ongoing

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #34637.

### CODE REVIEW

File: src/udev/udev-node.c [hunk 1]
```diff
@@ -588,10 +593,7 @@ int udev_node_remove(sd_device *dev) {
                 return log_device_debug_errno(dev, r, "Failed to get device path: %m");
 
         /* remove /dev/{block,char}/$major:$minor */
-        if (unlink(filename) < 0 && errno != ENOENT)
-                return log_device_debug_errno(dev, errno, "Failed to remove '%s': %m", filename);
-
-        return 0;
+        return node_remove_symlink(dev, filename);
```
Comment (bluca) [inline src/udev/udev-node.c:679]: this wasn't calling rmdir_parents before, is this intended?

**Review by bluca — COMMENTED** [review]
Thanks for the quick fix - given this will need to be backported, could you please rebase and do the refactors after the bug fix? So that we can minimize the backports

File: src/udev/udev-node.c (hunk 1 above)
Comment (yuwata) [reply src/udev/udev-node.c:679]: Yes. I do not think it is not necessary to keep /dev/block/ or /dev/char/ if empty.

**Comment by yuwata** [comment]
> Thanks for the quick fix - given this will need to be backported, could you please rebase and do the refactors after the bug fix? So that we can minimize the backports

OK. Cleanup commits are moved to the end.

**Comment by poettering** [comment]
hmm, is this really a proper fix for the issue? if the queue is clogged and we never clean up symlinks, then this might hit us in other cases too, no, on a "busy" system ("busy" in the sense of devices coming/going all the time, which might happen for synthetic devices such as loopback devices, or DM or so)?

**Comment by bluca** [comment]
> hmm, is this really a proper fix for the issue? if the queue is clogged and we never clean up symlinks, then this might hit us in other cases too, no, on a "busy" system ("busy" in the sense of devices coming/going all the time, which might happen for synthetic devices such as loopback devices, or DM or so)?

Even if there might be other sort of related issues, let's fix the one we actually hit first please. If more show up it can always be revisited

**Review by bluca — APPROVED** [review]
Can confirm it fixes the problem, thanks for the quick fix, will start the stable backports too

**Comment by poettering** [comment]
Umm? Can we maybe discuss this properly first? Why was this merged so quickly?

File: src/udev/udev-node.c [hunk 2]
```diff
@@ -423,6 +423,85 @@ static int node_get_current(const char *slink, int dirfd, char **ret_id, int *re
...
+
+        fname = path_startswith(slink, "/dev/disk/by-diskseq");
+        if (isempty(fname))
+                return 0;
+
+        if (device_is_devtype(dev, "partition")) {
+                _cleanup_free_ char *suffix = NULL;
+                const char *partn, *p;
+
+                /* Check if the symlink has an expected suffix "-part%n". See 60-persistent-storage.rules. */
```
Comment (poettering) [inline src/udev/udev-node.c:456]: i really don't like we have C code that special cases one udev rule we ship

File: src/udev/udev-node.c [hunk 3]
```diff
@@ -423,6 +423,85 @@ static int node_get_current(const char *slink, int dirfd, char **ret_id, int *re
...
+        _cleanup_free_ char *buf = NULL;
+        const char *fname, *diskseq;
+        int r;
+
+        assert(dev);
+        assert(slink);
+
+        if (!device_in_subsystem(dev, "block"))
+                return 0;
+
```
Comment (poettering) [inline src/udev/udev-node.c:447]: there really should be a comment here, explaining what this function does, this is so not clear to me

**Comment by bluca** [comment]
> Umm? Can we maybe discuss this properly first? Why was this merged so quickly?

6NfmQ

**Comment by poettering** [comment]
i still think this doesn't look like the right fix to me. if we stop processing symlink removal as long as udev is busy, then this will always create problems for us, afaics. It's just a matter of having enough symlinks in place that will change on each uevent, no? 

This is trivially going to be the case for /dev/disk/by-loop-inode/* and /dev/disk/by-loop-ref/*. The former always, and the latter if people have build tools that specify a randomized ref on each loopback invocation (which is a good thing, to have a stable identifier, that is unlikely to collide).

hence, i doubt this really the right approach, it works around the problem in one specific case, by encoding in C side information from udev rules, and that really should not be necessary. 

maybe a better fix would be to force symlink stack cleanup in some cases (i.e. every 500 udev jobs executed we'll "synthetically" drain the queue for a moment so that we can run the cleanup).

**Comment by DaanDeMeyer** [comment]
> > Umm? Can we maybe discuss this properly first? Why was this merged so quickly?
> 
> 6NfmQ

Downstream patches are a thing you know. If your build system is so insane that you can't carry a temporary downstream patch that's a pretty bad situation to be in

**Comment by poettering** [comment]
> > Umm? Can we maybe discuss this properly first? Why was this merged so quickly?
> 
> 6NfmQ

I doubt this PR is going to fix your problem, because of /dev/disk/by-loop-inode/.

If you want to backport half-reviewed stuff to your downstream, that's entirely fine by me, but why hurry this in upstream?

**Comment by bluca** [comment]
> I doubt this PR is going to fix your problem, because of /dev/disk/by-loop-inode/.

And yet it does, and it's trivial to see thanks to the reproducer _and_ the test yu added

> If you want to backport half-reviewed stuff to your downstream, that's entirely fine by me, but why hurry this in upstream?

To avoid the usual pitfall of blocking a fix for a real, actual bug, breaking real things (if I see it, others will too), because of imaginary ones that never actually happened. Let's stop doing that. If you want further changes that's fine, you can do them on top of the fix for the real issue, with the test case for the real bug included so that there are no regressions. Fixing bugs comes first, theorycrafting about hypothetical stuff comes after. Both can happen, in the right order.

**Comment by poettering** [comment]
IT'S NOT A PROPER FIX. IT'S A FIRST SUGGESTION OF A POSSIBLE FIX. And concerns where voiced which you ignored.

Why merge this? I don't get it? It's almost certainly not the right fix, and it's not what distros should backport.

**Comment by poettering** [comment]
it's fine if you do hacks like this on the "stable" branch, but please not in git main. git main will only hit real systems in a few months, we do not need to hurry there, and we can focus on correctness. 

BTW, if you has asked for a better thing to backport: just disable the by-inode + by-diskseq rules, and turn off its use in the image dissect code, and you should be good.

**Comment by bluca** [comment]
No, it is the right fix, as that directory pre-dates diskseq stuff, which was added on later, and it seems the combination just flew under the radar and nobody noticed this side effect - which is fine, it happens. It makes no sense to create this avalanche of symlinks, they serve no purpose, and as it turns out, can cause inode exhaustion which _completely break otherwise working systems_, so yes, fixing this is correct. If you want to then consider other theoretical cases like mapper devices - go for it, no problem whatsoever. But it's a separate problem.

**Comment by poettering** [comment]
> that directory pre-dates diskseq stuff

0d08db7f89ee665a9dcb6dd66c1f9e203192e8ec predates 1055172804e660df2e1c498dafa998ffd65e0c88

i.e. the rework of the cleanup logic was added *after* the diskseq symlink was added.

And noone disagrees that the issue should be fixed. We just disagree that suddenly after 2y a hotfix that people already voiced some concerns with gets hurried into our tree, without further discussion.

Sorry, @bluca please stop this. Do not hurry patches in anymore without proper discussion. Not into git main. Thank you.

**Comment by bluca** [comment]
> > that directory pre-dates diskseq stuff
> 
> [0d08db7](https://github.com/systemd/systemd/commit/0d08db7f89ee665a9dcb6dd66c1f9e203192e8ec) predates [1055172](https://github.com/systemd/systemd/commit/1055172804e660df2e1c498dafa998ffd65e0c88)
> 
> i.e. the rework of the cleanup logic was added _after_ the diskseq symlink was added.

/run/udev/links/ has existed for much longer than that one refactor

> And noone disagrees that the issue should be fixed. We just disagree that suddenly after 2y a hotfix that people already voiced some concerns with gets hurried into our tree, without further discussion.
> 
> Sorry, @bluca please stop this. Do not hurry patches in anymore without proper discussion. Not into git main. Thank you.

This scope creep happens way too often, and we need to stop it, because it's A) a waste of time and B) incredibly frustrating for contributors. It's fine, great even that seeing some bug fixes gives someone ideas about other maybe-related-maybe-not potential things that could also be done. But these need to be handled separately, and on top of fixing the original issue. We really, really, really need to stop blocking useful and needed fixees because of "while you are there, also X Y and Z", which is what's happening here. If you think the queue handling logic should be refactored, that's ok - nothing against it, let's do that, send a PR and I'll happily review and test it. It's ok to do things incrementally! But please let's stop trying to block useful stuff like this, it's immensely frustrating.

**Pull request merged**